### ISSUE TITLE
assets: garbage collection

### ISSUE THREAD

**Issue by eladb** [awslabs/aws-cdk#1332]
Assets which are uploaded to the toolkit's S3 bucket are never deleted. This will incur costs for users in the long term. We should come up with a story on how those should be garbage collected safely.

ECR assets today are tightly bound to their stack, but that's also not a good solution because it doesn't allow reusing ECR images across stacks in the account (which may be a huge time/space saver).

**Comment by sam-goodwin** [comment]
How do we generate keys when uploading? Just random?

If it were deterministic, we could use lifecycle rules of object versions: https://aws.amazon.com/about-aws/whats-new/2014/05/20/amazon-s3-now-supports-lifecycle-rules-for-versioning/

That way objects are only considered for deletion if they are not 'current'.

**Comment by sam-goodwin** [comment]
Now that I think about it, versioning would only help if we were regularly re-uploading files. That's probably not the case?

**Comment by eladb** [comment]
Life cycle rules sounds like a good option for sure. The object keys are based on the hash of the contents of the asset, so to avoid uploading in case the content hasn't changed ([code](https://github.com/awslabs/aws-cdk/blob/master/packages/aws-cdk/lib/api/toolkit-info.ts#L53))

**Comment by rix0rrr** [comment]
I'm not sure this works if there are two stacks, and only one is deployed with a new version of the asset. 

In my mind, the other stack should still point to the old version of the asset (should not be automatically updated), but now the asset will be aged out after a while and the undeployed stack will magically break after a certain time.

-----

Alternative idea (but much more involved and potentially not Free Tier): a Lambda that runs ever N hours which enumerates all stacks, detects the assets that are in use (from stack parameters or in some other way) and clears out the rest?

**Comment by rix0rrr** [comment]
> Now that I think about it, versioning would only help if we were regularly re-uploading files. That's probably not the case?
I thought you meant expiration for non-current versions. The latest version still stays alive indefinitely.

But I think this runs afoul of reuse across stacks.

**Comment by eladb** [comment]
:+1: on garbage collecting lambda that runs every week or so, with ability to opt out and some cost warnings on docs and online

**Comment by sam-goodwin** [comment]
Seems risky. What if a deployment happen while crawling?

**Comment by eladb** [comment]
Yeah perhaps only collect old assets (month old) and we can salt the object key such that if a whole month had passed, it will be a new object? 

Thinking out loud.... requires a design
